### ISSUE TITLE
rust: rustc segfaults while building crate encoding_c

### ISSUE THREAD

**Issue by Gottox** [voidlinux/void-packages#12503]
### System

* xuname:  
  Void 4.15.6_2 x86_64 GenuineIntel uptodate rrrrmd
* package:  
  rust-1.24.0_1

### Expected behavior
rustc should build the crate encoding_c

### Actual behavior
Segfault of rustc during build.

### Steps to reproduce the behavior
```
xbps-install cargo
git clone https://github.com/hsivonen/encoding_c
cd encoding_c
cargo build --verbose
```

The stacktrace of the segfaulting call: https://gist.github.com/Gottox/9e3d2e8ab32c3dc16036bf4cba0f342b

rustc from rustup works in both versions, 1.24.0 and 1.24.1:
```
rustc --version
rustc 1.24.1 (d3ae9a9e0 2018-02-27)
rustc --version
rustc 1.24.0 (4d90ac38c 2018-02-12)
```

**Comment by ebfe** [comment]
> #0 __strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:62

hmm, old cpu?

**Comment by Gottox** [comment]
calling rustc with `-C debuginfo=0` avoids the error.

**Comment by pullmoll** [comment]
FWIW segfault happens on my box as well
```
cat /proc/cpuinfo | grep model | head -n2
model		: 1
model name	: AMD Ryzen Threadripper 1950X 16-Core Processor
```

**Comment by pullmoll** [comment]
Added the following patch as `srcpkgs/rust/patches/fix-nullptr-LinkageName.patch` and the resulting `rustc` does not crash anymore. Of course I don't know what I'm doing there, just avoiding a NULL pointer in favor of a fixed string "(NULL)".
```
--- rustc-1.24.0-src/src/rustllvm/RustWrapper.cpp       2018-02-12 19:51:18.000000000 +0100
+++ rustc-1.24.0-src/src/rustllvm/RustWrapper.cpp       2018-03-14 21:03:42.277533250 +0100
@@ -725,7 +725,7 @@
   return wrap(VarExpr);
 #else
   return wrap(Builder->createGlobalVariable(
-      unwrapDI<DIDescriptor>(Context), Name, LinkageName,
+      unwrapDI<DIDescriptor>(Context), Name, LinkageName ? LinkageName : "(NULL)",
       unwrapDI<DIFile>(File), LineNo, unwrapDI<DIType>(Ty), IsLocalToUnit,
       InitVal, unwrapDIPtr<MDNode>(Decl)));
 #endif
```

**Comment by jnbr** [comment]
relevant change where the null pointer comes from is this: 
https://github.com/rust-lang/rust/commit/500dc14b21e7d1654b66fe246f33cf7741bccf64#diff-613e6ba013da492e36210d8bc6206ecfL1673

but I'm not sure how to fix that best either.

**Comment by Gottox** [comment]
hmm... I still can't see why rustups compiler does not crash in these situations.

**Comment by Arzte** [comment]
Unable to reproduce with rustup cargo on nightly or stable.

<pre>╭─<span style="background-color:#1E2127"> </span><span style="background-color:#1E2127"><font color="#9BA2B1"> </font></span><span style="background-color:#9BA2B1"><font color="#1E2127"> 08:59:18 </font></span><span style="background-color:#519DEB"><font color="#9BA2B1"> </font></span><span style="background-color:#519DEB"><font color="#1E2127"></font></span><span style="background-color:#519DEB"> </span><span style="background-color:#519DEB"><font color="#1E2127">~/git </font></span><font color="#519DEB"></font>                                                                                                                                                                   <font color="#1E2127"></font><span style="background-color:#1E2127"><font color="#88B966"> </font></span><span style="background-color:#1E2127"> </span><font color="#1E2127"> 5341 </font><span style="background-color:#1E2127"><font color="#DEB468"> 98%  </font></span><span style="background-color:#1E2127"> </span>
╰─ <font color="#88B966">git</font> clone https://github.com/hsivonen/encoding_c
Cloning into &apos;encoding_c&apos;...
remote: Counting objects: 2090, done.
remote: Total 2090 (delta 0), reused 0 (delta 0), pack-reused 2090
Receiving objects: 100% (2090/2090), 1.28 MiB | 593.00 KiB/s, done.
Resolving deltas: 100% (1515/1515), done.
╭─<span style="background-color:#1E2127"> </span><span style="background-color:#1E2127"><font color="#9BA2B1"> </font></span><span style="background-color:#9BA2B1"><font color="#1E2127"> 08:59:32 </font></span><span style="background-color:#519DEB"><font color="#9BA2B1"> </font></span><span style="background-color:#519DEB"><font color="#1E2127"></font></span><span style="background-color:#519DEB"> </span><span style="background-color:#519DEB"><font color="#1E2127">~/git </font></span><font color="#519DEB"></font>                                                                                                                                                                   <font color="#1E2127"></font><span style="background-color:#1E2127"><font color="#88B966"> </font></span><span style="background-color:#1E2127"> </span><font color="#1E2127"> 5342 </font><span style="background-color:#1E2127"><font color="#DEB468"> 98%  </font></span><span style="background-color:#1E2127"> </span>
╰─ <font color="#88B966">cd</font> <u>encoding_c</u> 
╭─<span style="background-color:#1E2127"> </span><span style="background-color:#1E2127"><font color="#9BA2B1"> </font></span><span style="background-color:#9BA2B1"><font color="#1E2127"> 08:59:35 </font></span><span style="background-color:#519DEB"><font color="#9BA2B1"> </font></span><span style="background-color:#519DEB"><font color="#1E2127"></font></span><span style="background-color:#519DEB"> </span><span style="background-color:#519DEB"><font color="#1E2127">~/git/encoding_c </font></span><span style="background-color:#88B966"><font color="#519DEB"> </font></span><span style="background-color:#88B966"><font color="#1E2127"> </font></span><span style="background-color:#88B966"> </span><span style="background-color:#88B966"><font color="#1E2127"> 6d41b99  master  v0.8.0 </font></span><font color="#88B966"></font>                                                                                                                       <font color="#1E2127"></font><span style="background-color:#1E2127"><font color="#88B966"> </font></span><span style="background-color:#1E2127"> </span><font color="#1E2127"> 5343 </font><span style="background-color:#1E2127"><font color="#DEB468"> 98%  </font></span><span style="background-color:#1E2127"> </span>
╰─ <font color="#88B966">cargo</font> build --verbose
<font color="#88B966"><b>    Updating</b></font> registry `https://github.com/rust-lang/crates.io-index`
<font color="#88B966"><b> Downloading</b></font> encoding_rs v0.7.2                                                 
<font color="#88B966"><b>   Compiling</b></font> cfg-if v0.1.2                                                      
<font color="#88B966"><b>     Running</b></font> `rustc --crate-name cfg_if /home/benjamin/.cargo/registry/src/github.com-1ecc6299db9ec823/cfg-if-0.1.2/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=91e4915dff2798a5 -C extra-filename=-91e4915dff2798a5 --out-dir /home/benjamin/git/encoding_c/target/debug/deps -L dependency=/home/benjamin/git/encoding_c/target/debug/deps --cap-lints allow`
<font color="#88B966"><b>   Compiling</b></font> encoding_rs v0.7.2
<font color="#88B966"><b>     Running</b></font> `rustc --crate-name encoding_rs /home/benjamin/.cargo/registry/src/github.com-1ecc6299db9ec823/encoding_rs-0.7.2/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=0336aba8820cc631 -C extra-filename=-0336aba8820cc631 --out-dir /home/benjamin/git/encoding_c/target/debug/deps -L dependency=/home/benjamin/git/encoding_c/target/debug/deps --extern cfg_if=/home/benjamin/git/encoding_c/target/debug/deps/libcfg_if-91e4915dff2798a5.rlib --cap-lints allow`
<font color="#88B966"><b>   Compiling</b></font> encoding_c v0.8.0 (file:///home/benjamin/git/encoding_c)
<font color="#88B966"><b>     Running</b></font> `rustc --crate-name encoding_c src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=519ef7bd13498cd3 -C extra-filename=-519ef7bd13498cd3 --out-dir /home/benjamin/git/encoding_c/target/debug/deps -C incremental=/home/benjamin/git/encoding_c/target/debug/incremental -L dependency=/home/benjamin/git/encoding_c/target/debug/deps --extern encoding_rs=/home/benjamin/git/encoding_c/target/debug/deps/libencoding_rs-0336aba8820cc631.rlib`
<font color="#88B966"><b>    Finished</b></font> dev [unoptimized + debuginfo] target(s) in 10.79 secs
╭─<span style="background-color:#1E2127"> </span><span style="background-color:#1E2127"><font color="#9BA2B1"> </font></span><span style="background-color:#9BA2B1"><font color="#1E2127"> 09:00:02 </font></span><span style="background-color:#519DEB"><font color="#9BA2B1"> </font></span><span style="background-color:#519DEB"><font color="#1E2127"></font></span><span style="background-color:#519DEB"> </span><span style="background-color:#519DEB"><font color="#1E2127">~/git/encoding_c </font></span><span style="background-color:#88B966"><font color="#519DEB"> </font></span><span style="background-color:#88B966"><font color="#1E2127"> </font></span><span style="background-color:#88B966"> </span><span style="background-color:#88B966"><font color="#1E2127"> 6d41b99  master  v0.8.0 </font></span><font color="#88B966"></font>                                                                                                                       <font color="#1E2127"></font><span style="background-color:#1E2127"><font color="#88B966"> </font></span><span style="background-color:#1E2127"> </span><font color="#1E2127"> 5344 </font><span style="background-color:#1E2127"><font color="#DEB468"> 98%  </font></span><span style="background-color:#1E2127"> </span>
╰─ <font color="#88B966">rustup</font> run stable cargo build --verbose
<font color="#88B966"><b>   Compiling</b></font> cfg-if v0.1.2
<font color="#88B966"><b>     Running</b></font> `rustc --crate-name cfg_if /home/benjamin/.cargo/registry/src/github.com-1ecc6299db9ec823/cfg-if-0.1.2/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=e1c5576e0db294e1 -C extra-filename=-e1c5576e0db294e1 --out-dir /home/benjamin/git/encoding_c/target/debug/deps -L dependency=/home/benjamin/git/encoding_c/target/debug/deps --cap-lints allow`
<font color="#88B966"><b>   Compiling</b></font> encoding_rs v0.7.2
<font color="#88B966"><b>     Running</b></font> `rustc --crate-name encoding_rs /home/benjamin/.cargo/registry/src/github.com-1ecc6299db9ec823/encoding_rs-0.7.2/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=2f3248c706731fae -C extra-filename=-2f3248c706731fae --out-dir /home/benjamin/git/encoding_c/target/debug/deps -L dependency=/home/benjamin/git/encoding_c/target/debug/deps --extern cfg_if=/home/benjamin/git/encoding_c/target/debug/deps/libcfg_if-e1c5576e0db294e1.rlib --cap-lints allow`
<font color="#88B966"><b>   Compiling</b></font> encoding_c v0.8.0 (file:///home/benjamin/git/encoding_c)
<font color="#88B966"><b>     Running</b></font> `rustc --crate-name encoding_c src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=225dbcf177abe113 -C extra-filename=-225dbcf177abe113 --out-dir /home/benjamin/git/encoding_c/target/debug/deps -C incremental=/home/benjamin/git/encoding_c/target/debug/incremental -L dependency=/home/benjamin/git/encoding_c/target/debug/deps --extern encoding_rs=/home/benjamin/git/encoding_c/target/debug/deps/libencoding_rs-2f3248c706731fae.rlib`
<font color="#88B966"><b>    Finished</b></font> dev [unoptimized + debuginfo] target(s) in 11.11 secs
</pre>

**Comment by jnbr** [comment]
It also does not crash, if rust is built with the in-tree llvm.

**Comment by pullmoll** [comment]
The reason is probably `/usr/include/llvm/ADT/StringRef.h` of llvm3.9

In llvm6.0.0 there is an inline function for constructing a string ref from a cstring that checks for `Str` being non-NULL and only then calls `::strlen(Str)`.

In llvm3.9-3.9.1 the corresponding inline function has an `assert(Str && "StringRef cannot be built from a NULL argument");` and in release mode makes ::strlen(NULL) crash.

**Comment by Gottox** [comment]
Yea, just digging through the sources. So rust should not be built against llvm3.9, but 4.0.1 upstream as does upstream.

**Comment by pullmoll** [comment]
We have also `julia` and `ponyc` depending on llvm3.9, so I think we should fix llvm3.9 anyway.

**Comment by Gottox** [comment]
Julia has similiar problems with llvm as rust: https://bugs.archlinux.org/task/57387

**Comment by Gottox** [comment]
Imho we should simply switch to internal llvms. @chneukirchen comments?

**Comment by chneukirchen** [comment]
If I don't need to compile it locally, I don't care. :)

**Comment by pullmoll** [comment]
BTW can someone update rust so it builds for x86_64-musl again?
The current distfile for x86_64-musl requires the previous libressl.

**Comment by Gottox** [comment]
fixed with a3991e750c1b966a95ced77b2e286042ae63ef63
